contributors/design-proposals/termination-reason.md

```diff
@@ -0,0 +1,317 @@
+time, propose to implement controller assisted propagation of termination 
+reasons from one API object to another (e.g Setting a termination reason for 
+a ReplicaSet which the controller will then propagate to its Pods). Such
+considerations will be deferred to future work.
+
+## Requirements
+
+1. Termination reason must be available to a container at the time its pre-stop 
+handler is invoked. In this way, applications can perform any actions
+necessary to configure the container's entry point prior to the entry point 
+receiving the `TERM` signal.
+1. The delivery mechanism for termination reason must be consumable by, and 
+ compatible with, existing HTTP and command based pre-stop handlers.
+1. Any headers, variables, or parameters used to communicate a termination 
+reason must not be hard coded, and the receiving container must be able to 
+override the defaults.
+
+## API Objects
+
+Termination reason is implemented as a generic string that can be associated 
+with the metadata of any object that supports graceful termination. Initially, 
+the reason will only be actionable with respect to Pods' containers, but this 
+generic delivery mechanism can be extended to other use cases.
+
+DeleteOptions is modified to carry a reason.
+
+```golang
+type DeleteOptions struct {
+	// Other fields omitted for brevity.
+	
+	// Reason indicates the reason for the deletion of an API Object that 
+	// undergo graceful termination upon deletion.
+	Reason string `json:"reason,omitempty"`
+}
+```
+
+The ObjectMeta struct is modified to carry the deletion reason via 
+an Object's metadata.
+
+```golang
+type ObjectMeta {
+	// Other fields omitted for brevity.
+	
+    // DeletionReason indicates the reason for the deletion of an API Object. 
+    // Its purpose is to provide an extra generic signal to watchers of API 
+    // Objects during the graceful termination process. 
+	DeletionReason string `json:"reason,omitempty"`
+}
+```
+
+DeleteExecAction and DeleteHttpAction are introduced to provide users with 
+the ability to configure the environment variable or HTTP Header that will be 
+used to convey a termination reason signaled by a DeletionReason.
+
+```golang
+// DeleteExecAction describes a "run in container" action that will be invoked
+// inside of a container prior to sending the TERM signal to the 
+// container's entry point.
+type DeleteExecAction struct {
+	// Command is the command line to execute inside the container, the working directory for the
+	// command  is root ('/') in the container's filesystem.  The command is simply exec'd, it is
+	// not run inside a shell, so traditional shell instructions ('|', etc) won't work.  To use
+	// a shell, you need to explicitly call out to that shell.
+	// +optional
+	Command []string
+	
+	// ReasonEnv is the environment variable that wil be populated with the 
+	// reason, if provided, for the Pod's termination. This variable defaults
+	// to "KUBE_DELETE_REASON"
+	ReasonEnv string
```

Comment:
There doesn't seem to be any way to support this. We are not able to set any environment variables once the container has started, and invoking the pre-stop handler as `env ReasonEnv=TerminationReason Command` will not work in containers that do not have `env`.

Reply:
Right, a running application cannot read environment variables afresh after start. I've implemented something close to this proposal using a CRD. In my case I signal the reason for termination by writing it into a file. Maybe this is also applicable to this proposal...